### ISSUE TITLE
use `argparse` for writing completion scripts

### ISSUE THREAD

**Issue by krader1961** [fish-shell/fish-shell#4226]
From a Gitter exchange earlier today:

@faho wrote

> I've been thinking about improving this using our shiny new argparse builtin (though the option-dialect that allows might not be exactly what other commands expect).

I wrote

> Using argparse should be correct for 98% of commands we want to generate completions for. Even those with two levels of subcommands. Because most commands use getopt()/getopt_long(). It obviously won't work for the oddball commands like find and tar which have their own idiosyncratic conventions for parsing CLI args. If and when we find sufficient need we can always add a new argparse flag to invoke different behavior. For example, to allow long options to be preceded by a single dash (as provided by "wgetopt_long_only()`).

> Also, even for typical commands we'll actually need an argparse flag that tells it to not treat flags which require an argument as an error. That way we can use it to cleanly see if the user just typed an argument requiring flag that needs a completion hint.

So I'm opening this issue to see if people think doing so offers clear advantages to the current, time tested, approach.

**Comment by faho** [comment]
I have locally rewritten the `git` completions to use it. It's really only useful in things like `__fish_git_needs_command` (which goes from 35 lines to 18) - even `__fish_git_stash_using_command` won't have a use it (because there are no options after `stash` but before the stash-subcommand).

What it could improve is our reactivity to options given on the commandline.

E.g. git has a `-C` option to run git as if $PWD were some other directory. We don't currently react to it. It would be nice if we could.

argparse could also be made a bit nicer for completions - in general, it should be more lenient (no errors, keeping an incomplete flag, assuming unknown options are booleans). 

One specific thing that might be nicer in general is the value of the flag variables for boolean options. Currently, that's the number of times the flag was given.

For e.g. `systemctl` completions, there is a "--user" boolean option that we want to pass to the completion script's `systemctl` calls. Currently, this requires something like

```fish
argparse "u-user" -- $argv
if set -q _flag_user
    set _flag_user --user
end
systemctl $_flag_user something
```

It would be nicer if the value was just the given options - `--user` here, but potentially also something equal to e.g. `set _flag_verbose -v --verbose -v`.

If someone is interested in the number of times an argument was given, they can use `count $_flag_user`.

**Comment by krader1961** [comment]
> It would be nicer if the value was just the given options - --user here, but potentially also something equal to e.g. set _flag_verbose -v --verbose -v.

That's an interesting idea. Certainly trivial to implement. And I can't see how it would break any existing uses since none of our functions I've converted to use `argparse` look at the value. Only whether the flag was seen at least once. And there should be zero external uses since it was added just a few days ago. I'll make that change ASAP.

> argparse could also be made a bit nicer for completions - in general, it should be more lenient (no errors, keeping an incomplete flag, assuming unknown options are booleans).

Agreed. At first glance "assuming unknown options are booleans" seems like a bad idea but doing that means we're more resilient when the completion script hasn't been "taught" about an available option. Perhaps because the user is using a version of the command that is newer than when the completions were written. Similarly, unknown long options should treated as if they had an option spec if they're of the form `--long=val` since, just as for unknown boolean flags, we should just assume it's a flag we haven't been taught to look for. The one thing we can't do is correctly handle unknown long flags if they're written like `--long val` since we don't know if `--long` is a bool or requires a value. But that's no worse that what completion scripts currently do and will almost certainly be better in practice.

Let me know if you think we're likely to see argparse used for completion scripts if that change is made. Since implementing that behavior is going to complicate the code and take a few days to implement I don't want to do the work unless we're pretty confident it will be used.

**Comment by faho** [comment]
>At first glance "assuming unknown options are booleans" seems like a bad idea but doing that means we're more resilient when the completion script hasn't been "taught" about an available option

That's the idea, yeah. It would also allow omitting any boolean options that we aren't interested in. E.g. for git we don't care about "--noglob-pathspecs".

>Let me know if you think we're likely to see argparse used for completion scripts if that change is made.

I'll try to look over the existing completion scripts to find any candidates. That might also give me some more ideas for improvement.

**Comment by faho** [comment]
>I'll try to look over the existing completion scripts to find any candidates.

A bunch of commands from the systemd project would benefit from this. They all have subcommands, and they have a bunch of options that we want to carry over for our invocation of the tool - stuff like the aforementioned "--user", but also "--host=" (which can be used to execute the operation on a remote host).

The only difference from `git` here is that the options can all appear before or after the subcommand, and options that aren't valid for a particular subcommand are ignored (e.g. `systemctl --message=foo list-units` just drops the "--message=foo").

That should actually work reasonably well with the current implementation.

I've been meaning to touch those scripts again anyway (there's some embarrassing stuff from my first contributions to fish in there I need to remove), so let me see how it goes!

**Comment by faho** [comment]
>I've been meaning to touch those scripts again anyway (there's some embarrassing stuff from my first contributions to fish in there I need to remove), so let me see how it goes!

So, I've tried `systemctl` now, and it's okay.

For the most part, the points I have are the same I've already made:

- I currently need to ignore argparse's stderr because it spews errors on unknown options and when the current token is an option that needs an argument.

- I can't use it to complete option arguments because it will have removed that option from the remaining argv, so I don't know which option it is

- Needing to find a short version of a long-only option is rather annoying here because it has a bunch of those (31 out of 42 total options). I have resorted to numbers (which are unused) and umlauts, especially since new options might be added to systemctl, and I don't want to have to change another option to fit it in. I suspect this would be even worse with e.g. `git`.

On the other hand, it's easy to react to the commandline, e.g. for things like `--state=something`.

So far, I have used a hybrid-style, where the options and option-arguments are completed via `complete`, and the rest is done via a function that uses `argparse`.

I'm just about ready to commit, which will also remove a bunch of function files.
